In [16]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/datasets/codemasterminds01/test-image/test-image/lion.jfif
/kaggle/input/datasets/codemasterminds01/90-animal-species-image-dataset/animals_dataset/beetle/687486f1cb.jpg
/kaggle/input/datasets/codemasterminds01/90-animal-species-image-dataset/animals_dataset/beetle/93fb82b04b.jpg
/kaggle/input/datasets/codemasterminds01/90-animal-species-image-dataset/animals_dataset/beetle/7d7f6c9750.jpg
/kaggle/input/datasets/codemasterminds01/90-animal-species-image-dataset/animals_dataset/beetle/0de750893f.jpg
/kaggle/input/datasets/codemasterminds01/90-animal-species-image-dataset/animals_dataset/beetle/2fecd5e844.jpg
/kaggle/input/datasets/codemasterminds01/90-animal-species-image-dataset/animals_dataset/beetle/514cd025fa.jpg
/kaggle/input/datasets/codemasterminds01/90-animal-species-image-dataset/animals_dataset/beetle/44931701fe.jpg
/kaggle/input/datasets/codemasterminds01/90-animal-species-image-dataset/animals_dataset/beetle/138f5c986e.jpg
/kaggle/input/datasets/codemasterminds0

KeyboardInterrupt: 

**Attach dataset:**

90 Animals Dataset

**Dataset structure must be:**

animals/

 ├── lion/
 
 ├── tiger/
 
 ├── dog/
 
 ├── cat/

 Each folder = one class

# STEP 2 — Import Required Libraries

In [2]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
import pickle

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D
from tensorflow.keras.layers import Flatten, Dense, Dropout
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.callbacks import EarlyStopping

2026-04-18 07:05:27.298088: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1776495927.494471      55 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1776495927.556814      55 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1776495928.026177      55 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1776495928.026217      55 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1776495928.026220      55 computation_placer.cc:177] computation placer alr

# STEP 3 — Define Dataset Path

(Change if dataset name differs)

In [3]:
dataset_path = "/kaggle/input/datasets/codemasterminds01/90-animal-species-image-dataset/animals_dataset"

# STEP 4 — Image Augmentation (Fix Overfitting)

This automatically creates extra training images

In [4]:
datagen = ImageDataGenerator(

    rescale=1./255,

    validation_split=0.2,

    rotation_range=20,

    width_shift_range=0.2,

    height_shift_range=0.2,

    zoom_range=0.2,

    horizontal_flip=True

)

# STEP 5 — Load Training Dataset

In [5]:
train_data = datagen.flow_from_directory(

    dataset_path,

    target_size=(128,128),

    batch_size=32,

    class_mode="categorical",

    subset="training"

)

Found 4320 images belonging to 90 classes.


# STEP 6 — Load Validation Dataset

Example output: 

Found 4320 images belonging to 90 classes

In [6]:
val_data = datagen.flow_from_directory(

    dataset_path,

    target_size=(128,128),

    batch_size=32,

    class_mode="categorical",

    subset="validation"

)

Found 1080 images belonging to 90 classes.


# STEP 7 — Build Improved CNN Model

This architecture prevents overfitting better than previous version.

In [7]:
# model = Sequential()

# model.add(Conv2D(32,(3,3),activation="relu",input_shape=(128,128,3)))
# model.add(MaxPooling2D(2,2))

# model.add(Conv2D(64,(3,3),activation="relu"))
# model.add(MaxPooling2D(2,2))

# model.add(Conv2D(128,(3,3),activation="relu"))
# model.add(MaxPooling2D(2,2))

# model.add(Flatten())

# model.add(Dense(128,activation="relu"))
# model.add(Dropout(0.5))

# model.add(Dense(train_data.num_classes,activation="softmax"))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
I0000 00:00:1776495954.881466      55 gpu_device.cc:2019] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 13757 MB memory:  -> device: 0, name: Tesla T4, pci bus id: 0000:00:04.0, compute capability: 7.5
I0000 00:00:1776495954.887598      55 gpu_device.cc:2019] Created device /job:localhost/replica:0/task:0/device:GPU:1 with 13757 MB memory:  -> device: 1, name: Tesla T4, pci bus id: 0000:00:05.0, compute capability: 7.5


Updated

In [8]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Input, Conv2D, MaxPooling2D
from tensorflow.keras.layers import Flatten, Dense, Dropout

model = Sequential([

    Input(shape=(128,128,3)),

    Conv2D(32,(3,3),activation="relu"),
    MaxPooling2D(2,2),

    Conv2D(64,(3,3),activation="relu"),
    MaxPooling2D(2,2),

    Conv2D(128,(3,3),activation="relu"),
    MaxPooling2D(2,2),

    Flatten(),

    Dense(128,activation="relu"),
    Dropout(0.5),

    Dense(train_data.num_classes,activation="softmax")

])

# STEP 8 — Compile Model

In [9]:
model.compile(

    optimizer="adam",

    loss="categorical_crossentropy",

    metrics=["accuracy"]

)

# STEP 9 — Add EarlyStopping (Stops Overfitting Automatically)

In [10]:
early_stop = EarlyStopping(

    monitor="val_loss",

    patience=5,

    restore_best_weights=True

)

# STEP 10 — Train CNN Model

Training stops automatically when validation stops improving.

In [11]:
history = model.fit(

    train_data,

    validation_data=val_data,

    epochs=25,

    callbacks=[early_stop]

)

/usr/local/lib/python3.12/dist-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


Epoch 1/25


I0000 00:00:1776495959.403411     139 service.cc:152] XLA service 0x7d35580084a0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
I0000 00:00:1776495959.403448     139 service.cc:160]   StreamExecutor device (0): Tesla T4, Compute Capability 7.5
I0000 00:00:1776495959.403451     139 service.cc:160]   StreamExecutor device (1): Tesla T4, Compute Capability 7.5
I0000 00:00:1776495959.822707     139 cuda_dnn.cc:529] Loaded cuDNN version 91002
2026-04-18 07:06:01.604787: E external/local_xla/xla/stream_executor/cuda/cuda_timer.cc:86] Delay kernel timed out: measured time has sub-optimal accuracy. There may be a missing warmup execution, please investigate in Nsight Systems.
2026-04-18 07:06:01.752718: E external/local_xla/xla/stream_executor/cuda/cuda_timer.cc:86] Delay kernel timed out: measured time has sub-optimal accuracy. There may be a missing warmup execution, please investigate in Nsight Systems.


  1/135 ━━━━━━━━━━━━━━━━━━━━ 14:30 6s/step - accuracy: 0.0000e+00 - loss: 4.5195

I0000 00:00:1776495963.867163     139 device_compiler.h:188] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.


135/135 ━━━━━━━━━━━━━━━━━━━━ 78s 533ms/step - accuracy: 0.0131 - loss: 4.5080 - val_accuracy: 0.0194 - val_loss: 4.4334
Epoch 2/25
135/135 ━━━━━━━━━━━━━━━━━━━━ 49s 364ms/step - accuracy: 0.0189 - loss: 4.4093 - val_accuracy: 0.0472 - val_loss: 4.3227
Epoch 3/25
135/135 ━━━━━━━━━━━━━━━━━━━━ 50s 371ms/step - accuracy: 0.0255 - loss: 4.3144 - val_accuracy: 0.0472 - val_loss: 4.2419
Epoch 4/25
135/135 ━━━━━━━━━━━━━━━━━━━━ 48s 358ms/step - accuracy: 0.0383 - loss: 4.2332 - val_accuracy: 0.0657 - val_loss: 4.1627
Epoch 5/25
135/135 ━━━━━━━━━━━━━━━━━━━━ 49s 367ms/step - accuracy: 0.0446 - loss: 4.1835 - val_accuracy: 0.0694 - val_loss: 4.0974
Epoch 6/25
135/135 ━━━━━━━━━━━━━━━━━━━━ 49s 362ms/step - accuracy: 0.0632 - loss: 4.1126 - val_accuracy: 0.0667 - val_loss: 4.0493
Epoch 7/25
135/135 ━━━━━━━━━━━━━━━━━━━━ 50s 372ms/step - accuracy: 0.0714 - loss: 4.0182 - val_accuracy: 0.0880 - val_loss: 3.9347
Epoch 8/25
135/135 ━━━━━━━━━━━━━━━━━━━━ 51s 376ms/step - accuracy: 0.0806 - loss: 3.9834 - val

# STEP 11 — Plot Accuracy Graph

Correct graph should look like:

Train ↑
Validation ↑

In [2]:
plt.plot(history.history["accuracy"])
plt.plot(history.history["val_accuracy"])

plt.title("Model Accuracy")

plt.legend(["Train","Validation"])

plt.show()

NameError: name 'plt' is not defined

# STEP 12 — Save Trained Model (Deployment File)

Output file:

animal_cnn_model.keras

In [13]:
model.save("animal_cnn_model.keras")

# STEP 13 — Save Class Labels

Very important for deployment

Output:

labels.pkl

In [14]:
labels = train_data.class_indices

pickle.dump(labels, open("labels.pkl","wb"))

# STEP 14 — Test Model on New Image

Example:

In [15]:
from tensorflow.keras.preprocessing import image

img = image.load_img(

    "/kaggle/input/datasets/codemasterminds01/test-image/test-image/lion.jfif",

    target_size=(128,128)

)

img_array = image.img_to_array(img)/255

img_array = np.expand_dims(img_array,axis=0)

prediction = model.predict(img_array)

predicted_class = np.argmax(prediction)

print(predicted_class)

1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 707ms/step
46
